test/test.lua

```diff
@@ -2,15 +2,60 @@ require 'cudnn'
 require 'cunn'
 
 local cudnntest = torch.TestSuite()
-local precision_forward = 1e-4
-local precision_backward = 1e-2
-local precision_jac = 1e-3
-local precision_io = 1e-5
 local nloop = 1
 local times = {}
 local mytester
 local jac = nn.Jacobian
 
+
+local testparams_half = {
+   test_type = 'torch.CudaHalfTensor',
+   precision_forward = 2e-1,
+   precision_backward = 3,
+   precision_jac = 1e-3,
+   precision_io = 1e-1,
+}
+
+local testparams_float = {
+   test_type = 'torch.CudaTensor',
+   precision_forward = 1e-4,
+   precision_backward = 1e-2,
+   precision_jac = 1e-3,
+   precision_io = 1e-5,
+}
+
+-- TODO: find out why the errors are so huge
+local testparams_double = {
+   test_type = 'torch.CudaDoubleTensor',
+   precision_forward = 1e+4,
```

Comment:
I have checked with caffe guys and they say there is no way those errors may have numeric origin (when they change float to double in their networks, it never increases the error).
So I think, it's reasonable to assume we may have some layout/alignment/copy error in templated types handling, at least for double type. My other segfault error in the test may mean the error is generic but does not manifest itself numerically in 32 bit and smaller types.